tests/test_report.py

```diff
@@ -0,0 +1,81 @@
+import json
+import numbers
+import os
+import sys
+
+import mock
+import pytest
+
+from iopipe.config import set_config
+from iopipe.report import Report
+
+from .mock_context import MockContext
+
+
+def assert_valid_schema(obj, schema=None, path=None):
+    """Asserts that an object matches the schema"""
+
+    if not schema:
+        schema_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'schema.json')
+        with open(schema_path) as f:
+            schema = json.loads(f.read())
+
+    if not path:
+        path = []
+
+    for key in obj:
+        if key not in schema:
+            continue
+
+        if isinstance(obj[key], dict):
+            assert_valid_schema(obj[key], schema[key], path + [key])
+        elif isinstance(obj[key], list):
+            for item in obj[key]:
+                assert_valid_schema(item, schema[key][0], path + [key])
+        elif schema[key] == 'b':
+            assert isinstance(obj[key], bool), '%s not a boolean' % '.'.join(path + [key])
+        elif schema[key] == 'i':
+            assert isinstance(obj[key], int), '%s not a integer' % '.'.join(path + [key])
+        elif schema[key] == 'n':
+            assert isinstance(obj[key], numbers.Number), '%s not a number' % '.'.join(path + [key])
+        elif schema[key] == 's':
+            assert isinstance(obj[key], str), '%s not a string' % '.'.join(path + [key])
+
+
+@mock.patch('iopipe.report.send_report')
+def test_report_linux_system(mock_send_report):
+    """Asserts that fields collected by the system module are present in a report"""
+
+    if not sys.platform.startswith('linux'):
+        pytest.skip('this test requires linux, skipping')
+
+    mock_context = MockContext()
+
+    report = Report(set_config(), mock_context)
+    report.send()
+
+    assert 'aws' in report.report
+
+    for key in ['functionName', 'functionVersion', 'awsRequestId',
```

Comment:
You should be able to use recursion to walk through this object, vs going through each key individually (perks include the tests failing when the schema has updated but the agent has not)

Reply:
The `assert_valid_Schema` already does this. These tests just assert those system-read fields are present to catch what was observed yesterday.

Reply:
That 'shelps thanks, will comment on that function. I think this test is redundant then/not needed.

Reply:
There are fields in the report that are not in the schema and there are fields in the schema that are not in the report. Myy initial plan was to make all the values that are required in the `schema.json` uppercase, so "N" instead of "n" for a required number. But since we want to centralize where the `schema.json` lives, this would also need to e implemented on the `iopipe/iopipe` side. So thinking we should circle back on this. I'm hesitant to remove any assertions that actually test something that was observed until we have more robust schema checking, ideally a way in the `schema.json` to denote whether or not it is required.

Unless we're ok with removing fields that aren't in the schema? Like `uptime` for example, which appears in both agents.

Reply:
It's ok to remove fields not in the schema, since they don't end up in the final data anyhow.

As to some fields not being required ... since it's a small number (performance_entries, custom_metrics, plugins..) handling it as a "unless it's these ones that are ok" is fine imo